### ISSUE TITLE
`System.AccessViolationException` caused by combination of `AssemblyBuilder`, `GC.TryStartNoGCRegion`, and `GC.Collect`

### ISSUE THREAD

**Issue by hamarb123** [dotnet/runtime#82746]
### Description

`System.AccessViolationException` is caused by the code below. The code basically uses `Parallel.For` to create a bunch of assemblies, and then collects the GC and prints its iteration occasionally. The assemblies have "randomised" methods that are created and invoked. You may need to adjust the constants to get it to happen quicker. The `GC.TryStartNoGCRegion` etc calls seem to be necessary for it to happen (at least in a timely manner). The `GC.Collect` etc calls seem to make the code run much, much faster, I don't know if they're necessary though.

The methods it generates are "random", but not very useful. I doubt they are specific to the problem, the code is just the first randomisable code that came to my mind. It's certainly not a useful calculation.

It also reproduces with `GCCollectionMode.Forced` or with (`allocation = 250_000_000`, `parallel_instances = 125`, and `parallel_repeats = 80`), but doesn't seem to with only `allocation = 250_000_000` (I only tried this 3x up to 160,000 - my guess is that it runs out of memory and causes a GC, this making it happen in the wrong spot to reproduce this).

### Reproduction Steps

Run the code with .NET 7 on Windows x64 with a few GB of RAM free for this process only. You may need to adjust the constants if you don't have ~10 effective cores, idk.

```csharp
using System.Reflection;
using System.Reflection.Emit;
using System.Runtime.InteropServices;

internal class Program
{
	const int allocation = 1_000_000_000; //how much memory to allocate for the no GC latency region
	const int parallel_instances = 500; //how many instances of MakeDynamicMethod to run in parallel in the Parallel.For
	const int parallel_repeats = 20; //how many times to repeat setting up the no GC latency region, running the Parallel.For, and then dismantling the no GC latency region and collecting the GC
	const int code_iterations = 4; //how many random sections of code to include

	static void Main(string[] args)
	{
		var options = new ParallelOptions() { MaxDegreeOfParallelism = -1 };

		//loop indefinitely with a counter
		for (int i = 0; ; i++)
		{
			//loop a bunch of times before printing and collecting 2 extra times
			for (int j = 0; j < parallel_repeats; j++)
			{
				//begin no latency region
				GC.TryStartNoGCRegion(allocation);

				//call MakeDynamicMethod a bunch of times
				Parallel.For(0, parallel_instances, options, (x) => MakeDynamicMethod());

				//end no latency region
				GC.EndNoGCRegion();

				//collect GC memory
				GC.Collect(GC.MaxGeneration, GCCollectionMode.Aggressive, true, true);
				GC.WaitForPendingFinalizers();
			}

			//collect GC memory again because why not
			GC.Collect(GC.MaxGeneration, GCCollectionMode.Aggressive, true, true);
			GC.WaitForPendingFinalizers();
			GC.Collect(GC.MaxGeneration, GCCollectionMode.Aggressive, true, true);
			GC.WaitForPendingFinalizers();

			//print iteration
			var idx = (i + 1) * parallel_instances * parallel_repeats;
			Console.WriteLine(idx.ToString("N0"));
		}
	}

	private static MethodInfo _toString = typeof(object).GetMethod("ToString")!;
	private static MethodInfo _getHashCode = typeof(object).GetMethod("GetHashCode")!;
	static void MakeDynamicMethod()
	{
		//create the assembly and the type
		AssemblyBuilder b = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("abc"), AssemblyBuilderAccess.RunAndCollect);
		var t = b.DefineDynamicModule("module").DefineType("def");

		//create the method
		var m = t.DefineMethod("method", MethodAttributes.Public | MethodAttributes.Static, typeof(int), null);

		//emit the random method instructions
		var ilGen = m.GetILGenerator();
		var l0 = ilGen.DeclareLocal(typeof(int));
		ilGen.Emit(OpCodes.Ldc_I4, Random.Shared.Next());
		for (int i = 0; i < code_iterations; i++)
		{
			if (Random.Shared.Next(2) == 0)
			{
				//equivalent to local = pop; push(local.ToString.GetHashCode());
				ilGen.Emit(OpCodes.Stloc, l0);
				ilGen.Emit(OpCodes.Ldloca, l0);
				ilGen.Emit(OpCodes.Constrained, typeof(int));
				ilGen.Emit(OpCodes.Callvirt, _toString);
				ilGen.Emit(OpCodes.Callvirt, _getHashCode);
			}
			else
			{
				//equivalent to local = pop; push((int)(nint)(nuint)(&local) * local + random constant);
				ilGen.Emit(OpCodes.Stloc, l0);
				ilGen.Emit(OpCodes.Ldloca, l0);
				ilGen.Emit(OpCodes.Conv_U);
				ilGen.Emit(OpCodes.Conv_I);
				ilGen.Emit(OpCodes.Conv_I4);
				ilGen.Emit(OpCodes.Ldloc_0);
				ilGen.Emit(OpCodes.Mul);
				ilGen.Emit(OpCodes.Ldc_I4, Random.Shared.Next());
				ilGen.Emit(OpCodes.Add);
			}
		}
		ilGen.Emit(OpCodes.Ret);

		//compile and call the method
		t.CreateType();
		var del = t.GetMethod("method")!.CreateDelegate<Func<int>>();
		del();
	}
}
```

### Expected behavior

Code keeps iterating indefinitely and achieves nothing of value.

### Actual behavior

Code crashes sometimes randomly with `Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.`. I generally see this by 160,000, you may have to run it a few times before it crashes.

### Regression?

Doesn't seem to happen in .NET 6 (using `GCCollectionMode.Forced` instead). I only tried it 3x though up to 160,000.

### Configuration

.NET 7.0.3
This is running in Windows 10.0.19045.2673 x64 Parallels VM with 10 "processors" and 12GB RAM.
Doesn't seem to reproduce on the machine that the VM is running on: macOS 12.6.3 x64 with .NET 7.0.2 (have to set `allocation` to `250_000_000` though to get it to run, also set `parallel_instances = 125` and `parallel_repeats = 80`) which has 8 cores with hyperthreading.
It would be good if someone else with a x64 Windows machine could reproduce it. It would also be good to see if it happens with other architectures and OSes.

### Other information

It's probably related to the GC. I think the timing of the GC collection is important to cause this. I have no idea if this is a duplicate of something 😆.

**Comment by hamarb123** [comment]
The area should probably not be reflection-emit, but rather GC

**Comment by mangod9** [comment]
Thanks for reporting the issue @hamarb123. Could you please check if it reproduces after setting `DOTNET_GCName=clrgc.dll`. This will help narrow down whether the problem is related to the new GC implementation. Thanks. 

@cshung does this look related to the latest fixes you have made in `7.0.4`?

**Comment by hamarb123** [comment]
I used `$env:DOTNET_GCName = 'clrgc.dll'` in powershell (is that the right way of doing it in powershell?)
I can still reproduce it with that set.

**Comment by janvorli** [comment]
It sounds likely that it would have the same underlying problem that I've fixed in #82841 5 days ago.
@hamarb123 could you please test it with DOTNET_EnableWriteXorExecute env variable set to 0 ($env:DOTNET_EnableWriteXorExecute=0 in powershell)?

If it doesn't repro with that setting, my bet is that it is very likely that issue.

**Comment by hamarb123** [comment]
Hi, sorry for the delay in response, it indeed doesn't repro with that setting. Thanks.

**Comment by hamarb123** [comment]
Would that fix have made it into .NET 8 preview 2? Let me know, and if so I can give it a try with that to confirm it is definitely fixed.

**Comment by mangod9** [comment]
The fix should be included in preview 3
